cache/contenthash/path.go

```diff
@@ -1,108 +1,111 @@
+// This code mostly comes from <https://github.com/cyphar/filepath-securejoin>.
+
+// Copyright (C) 2014-2015 Docker Inc & Go Authors. All rights reserved.
```

Comment:
Not a fan of these new copyrights(link to reference repository is completely fine). Afaics (and based on running the new test with old implementation) the previous implementation of `rootPath` does not need to actually change to fix the issue and it already evaluates per path component.

Otherwise, I don't really mind if it is similar to one implementation or another.

Reply:
Can we move this to a comment rather than a copyright header, @cyphar? Although it is heavily influenced by the linked source and should be directly attributed, some structural changes, plus the existing file history, make it not quite a copy-and-modify.

Approaches in the project so far:
- [inline](https://github.com/moby/buildkit/blob/4cf5e3405210413b726e0518847fe05776b9bcfe/util/overlay/overlay_linux.go#L364)
- [copy and modify](https://github.com/moby/buildkit/blob/master/util/tracing/otlptracegrpc/client.go)

Reply:
I feel like I'm missing something -- what is the difference between those examples and the copyright header in this file? The new code in `path.go` is a strict copy-and-modify from `SecureJoinVFS` with the addition of the callback. If it were from a different author, I would've credited it the same way...

The main reason for switching away from the old implementation is that it was far more complicated than necessary and (imho, given the mistakes made with the other recursive implementations) only happened to produce the correct results. And unless the loop in `rootPath` is unnecessary, I also suspect a straightforward iterative implementation is more efficient anyhow.

(As for including copyright headers in general -- if you'd like me to remove the copyright notice and obscure the copyrights by hiding it behind a link, I don't really care, but the legal status is the same regardless. buildkit doesn't have a CLA, after all. All of my other changes to the repo are also copyrighted by me, as is the case for every other contributor.)

Reply:
🙏 Thanks for the explanation, @cyphar; that helped clear things up. Drops into the copy-and-mod case. No issues from me ;)

Reply:
(personal opinion)
This isn't about us protecting against something or trying to obfuscate. To me it is about when someone looks at buildkit codebase I don't want them to be confused about what file they can use or modify and what these scary headers mean. The only thing they should care about that all codebase is under a very liberal Apache license. This isn't the most important thing, but the simpler we can make this, the better. In this case, we have obvious alternatives as a) you are the author of the referenced code that is being brought in, and b) we could just keep the old code without breaking any buildkit functionality.

> is that it was far more complicated than necessary

Not sure if I fully agree with that. I think in the old code it is easier to follow what is happening as it doesn't have this big loop with many early return and skip conditions.

Reply:
I don't like arguing about licenses, but the code was based on some Docker code which was in turn based on some Go stdlib code. Given that, while I did basically rewrite it it's probably not okay for me to remove the copyright header. (Also, removing the copyright header if I was the sole copyright holder would probably be interpreted as re-licensing the code.)

On the topic of scariness, I think hiding the copyright status makes lives harder for folks who actually care about those problems (when dealing with license compatibility with SUSE lawyers, they really don't like obfuscated license terms). Folks that care enough about licenses to bother reading the copyright header would understand that the BSD-like license used by the code is Apache-2.0 compatible.

<hr>

As for the old implementation, it actually does have an issue because of the use of `filepath.Join`. If you encounter a symlink with the contents `a/../b` it will be resolved to `b` even if `a` is a symlink -- this is incorrect. I can write a test for this case if you really like. To solve this and keep the recursive implementation we would be to iteratively add the components, and then recursively do `walkLinks` for each one -- this would be far less efficient because you would be re-doing most of the lookup for each new component.

FWIW even if this wasn't an issue, I don't see how the `if i == linksWalked` loop combined with the recursive implementation of an iterative algorithm is "easier to follow". The iterative version is shorter (especially if you exclude comments) and the loop only has three `continue`/`break` conditions (one is an optimisation to avoid re-evaluating the same path multiple times, one is the implementation of `followTrailing` -- which might not be necessary, and the third is the standard "add a component" case).

Reply:
I've pushed some tests for `rootPath` which fail with the old implementation:

```
--- FAIL: TestRootPathSymlinks (0.00s)
    --- FAIL: TestRootPathSymlinks/resolve("link2/target") (0.00s)
        path_test.go:71:
                Error Trace:    /home/cyphar/.local/src/github.com/moby/buildkit/cache/contenthash/path_test.go:71
                Error:          Not equal:
                                expected: "/tmp/TestRootPathSymlinks3433359532/001/target"
                                actual  : "/tmp/TestRootPathSymlinks3433359532/001/link2/final"

                                Diff:
                                --- Expected
                                +++ Actual
                                @@ -1 +1 @@
                                -/tmp/TestRootPathSymlinks3433359532/001/target
                                +/tmp/TestRootPathSymlinks3433359532/001/link2/final
                Test:           TestRootPathSymlinks/resolve("link2/target")
    --- FAIL: TestRootPathSymlinks/resolve("link2/notaloop") (0.00s)
        path_test.go:70:
                Error Trace:    /home/cyphar/.local/src/github.com/moby/buildkit/cache/contenthash/path_test.go:70
                Error:          Received unexpected error:
                                too many links
                                github.com/moby/buildkit/cache/contenthash.init
                                        /home/cyphar/.local/src/github.com/moby/buildkit/cache/contenthash/checksum.go:27
                                runtime.doInit1
                                        /usr/lib64/go/1.21/src/runtime/proc.go:6735
                                runtime.doInit
                                        /usr/lib64/go/1.21/src/runtime/proc.go:6702
                                runtime.main
                                        /usr/lib64/go/1.21/src/runtime/proc.go:249
                                runtime.goexit
                                        /usr/lib64/go/1.21/src/runtime/asm_amd64.s:1650
                Test:           TestRootPathSymlinks/resolve("link2/notaloop")
FAIL
FAIL    github.com/moby/buildkit/cache/contenthash      0.008s
FAIL
```